pkg/certificate/mrc_reconciler.go

```diff
@@ -1,214 +1,228 @@
 package certificate
 
 import (
-	"time"
-
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-
 	"github.com/openservicemesh/osm/pkg/apis/config/v1alpha2"
-	"github.com/openservicemesh/osm/pkg/constants"
+	"github.com/openservicemesh/osm/pkg/errcode"
 )
 
-// UseCase is how the certificate is used. Each use case will have its own rotation mechanism, that is necessary to
-// determine if we can propagate the rotation forward.
-type UseCase string
+func (m *Manager) handleMRCEvent(event MRCEvent) error {
+	log.Debug().Msgf("handling MRC event for MRC %s", event.MRCName)
+	// TODO(#5226): optimize event handling to reduce cost of listing all MRCs for each event
```

Comment:
Note that we could simply keep the MRC on the issuer struct, and then we have the 2 mrc's we care about. This seems like a simple enough change?

Reply:
When I chatted with @jaellio last week, it got a bit complicated on the edge cases I believe.  We've got https://github.com/openservicemesh/osm/issues/5226 to track.  If we want this in before release we can work can follow up there, but I would be in favor of doing that separately so we can unblock the other PRs that are waiting on this.

Reply:
Ok, I won't block, but I'll point it that it looks like we could do a drop in replacement into this code:

```
...
   desiredSigning, desiredValidating := m.getSigningAndValidating(event.MRC)
   // note: no need to filter out inactive anymore, or call ListMeshRootCertificates
...

func getSigningAndValidating(newMRC MRC) {
  mrcs := make(map[string]MRC)
  mrcs[m.signingIssuer.ID] = m.signingIssuer.mrc
  mrcs[m.validatingIssuer.ID] = m.validatingIssuer.mrc
  mrcs[newMRC.Name] = newMRC
  if len(mrcs) > 2 {.../*err*/ }
 
   ... // continue as normal
}
```

I think using a map like this would also clean up some of the logic in that method. But  again, non-blocking

Reply:
An example of a case where this gets tricky is when a controller restarts. In that case the existing manager won't have any sign/validating issuers set.  By approaching it this way we are handling the events in a trigger level fashion.

This means that on the first event that comes through the issuer/validator would be the same but this might not be the case, there could be more than one Active/Passive.  In this case there are some conditions where the controller could start to re-issue certificates since the `ShouldRotate` function would see this as a change. Once more events came through it would be normalize but this could be a spot where we have strange behavior intermittently.

Using a custom indexer (https://github.com/openservicemesh/osm/issues/5226) that only contains active MRCs we could be more level triggered and  solve this problem of having to filter and list all the MRCS.  

Maybe we could combine the two options as well, which would allow us to exit earlier in the case things are set?

Reply:
non-blocking at this point